librad/src/git/storage.rs

```diff
@@ -0,0 +1,371 @@
+        let repo = git2::Repository::init_opts(
+            paths.git_dir(),
+            git2::RepositoryInitOptions::new()
+                .bare(true)
+                .no_reinit(true)
+                .external_template(false),
+        )?;
+
+        let mut config = repo.config()?;
+        config.set_str("user.name", "radicle")?;
+        config.set_str("user.email", &format!("radicle@{}", PeerId::from(&key)))?;
+
+        Ok(Self {
+            backend: Arc::new(Mutex::new(repo)),
+            key,
+        })
+    }
+
+    pub fn create_repo<T>(self, meta: &Entity<T>) -> Result<Repo, repo::Error>
+    where
+        T: Serialize + DeserializeOwned + Clone + Default,
+        EntityData<T>: EntityBuilder,
+    {
+        Repo::create(self, meta)
+    }
+
+    pub fn open_repo(self, urn: RadUrn) -> Result<Repo, repo::Error> {
+        Repo::open(self, urn)
+    }
+
+    pub fn clone_repo<T>(self, url: RadUrl) -> Result<Repo, repo::Error>
+    where
+        T: Serialize + DeserializeOwned + Clone + Default,
+        EntityData<T>: EntityBuilder,
+    {
+        Repo::clone(self, url)
+    }
+
+    // Utils
+
+    pub(super) fn lock(&self) -> MutexGuard<git2::Repository> {
+        self.backend.lock().unwrap()
+    }
+
+    pub(crate) fn has_commit(&self, urn: &RadUrn, oid: git2::Oid) -> Result<bool, Error> {
+        let span = tracing::warn_span!("Storage::has_commit", urn = %urn, oid = %oid);
+        let _guard = span.enter();
+
+        if oid.is_zero() {
+            return Ok(false);
+        }
+
+        let git = self.lock();
+        let commit = git.find_commit(oid);
+        match commit {
+            Err(e) if is_not_found_err(&e) => {
+                tracing::warn!("commit not found");
+                Ok(false)
+            },
+            Ok(commit) => {
+                let namespace = &urn.id;
+                let branch = urn.path.deref_or_default();
+                let branch = branch.strip_prefix("refs/").unwrap_or(branch);
+
+                let refs = References::from_globs(
+                    &git,
+                    &[format!("refs/namespaces/{}/refs/{}", namespace, branch)],
+                )?;
+
+                for (_, oid) in refs.peeled() {
+                    if oid == commit.id() || git.graph_descendant_of(oid, commit.id())? {
+                        return Ok(true);
+                    }
+                }
+
+                Ok(false)
+            },
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    pub(crate) fn has_ref(&self, reference: &Reference) -> Result<bool, Error> {
+        self.lock()
+            .find_reference(&reference.to_string())
+            .map(|_| true)
+            .or_else(|e| {
+                if is_not_found_err(&e) {
+                    Ok(false)
+                } else {
+                    Err(e.into())
+                }
+            })
+    }
+
+    pub(crate) fn has_urn(&self, urn: &RadUrn) -> Result<bool, Error> {
```

Comment:
Could clean up `has_urn` and `has_ref` since they're essentially the same other than how to make the `str`.

Reply:
Ya maybe... since we don't know the `RefsCategory`, we'd need an `impl FromStr for Reference`. But then, this may also be too strict.

A useful thing would be some combinator for this `if is_not_found_err` business, but I couldn't figure out what the type of it should be 🤔

Reply:
Ya, I saw that around the place :thinking: Maybe something like:
```rust
fn on_not_found<T, F>(e: git2::Error, f: F) -> Result<T, Error> where FnOnce() -> T {
  if is_not_found(e) {
    Ok(f())
  } else {
    Err(e.into())
  }
}
```

Reply:
Ya.. although that's not a lot nicer, because we can't partially apply functions in Rrrrrust. Ideally, the receiver should be `Result<T, git2::Error>`, so you can use it like `git2_thing.map_not_found(|| bla())`.

Reply:
Ohhh:

```rust
pub trait Git2ErrorExt<T> {
    fn map_not_found<E: From<git2::Error>>(self, f: impl FnOnce() -> Result<T, E>) -> Result<T, E>;
}

impl<T> Git2ErrorExt<T> for Result<T, git2::Error> {
    fn map_not_found<E: From<git2::Error>>(self, f: impl FnOnce() -> Result<T, E>) -> Result<T, E> {
        self.or_else(|e| {
            if is_not_found_err(&e) {
                f()
            } else {
                Err(e.into())
            }
        })
    }
}
```

Reply:
:kissing_closed_eyes::ok_hand:

Reply:
2aba2b816aea79d29adefe94cff03760a1971b7b

"Wat?" moment: I had to change this to not use `impl Trait`, but proper trait bounds, because:

If type inference fails, one is not allowed to use the turbofish when there is an `impl Trait` involved in _one_ of the arguments (I actually needed to annotate the type of `E`, but no luck).